README.md
# XLabs

[![patreon](https://img.shields.io/badge/patreon-support-blue.svg?logo=patreon)](https://www.patreon.com/xlabsproject)
[![discord](https://img.shields.io/endpoint?url=https://momo5502.com/iw4x/members-badge.php)](https://discord.gg/sKeVmR3)
[![website](https://img.shields.io/badge/Repackers-_Website-blue)](https://rimmyscorner.com/)

<p align="center">
  <img src="assets/github/banner-iw4x.png?raw=true" />
</p>

<br/>

<p align="center">
  <img src="assets/github/banner-s1x.png?raw=true" />
</p>

<br/>

<p align="center">
  <img alig src="assets/github/banner-iw6x.png?raw=true"/>
</p>

## Prerequisites: Installing Git

If you don't have `git` installed on your machine, follow these steps:

1. Visit the [Git download page](https://git-scm.com/downloads).
2. Download the appropriate version for your operating system.
3. Run the installer.
   - During the installation, you will be asked if you want to add Git to your system's PATH environment variable. Ensure you choose the option to do so. This allows you to use Git from the command line without specifying its full path.
4. After installation, open a terminal or command prompt and type `git --version` to ensure Git is correctly installed.

## Download
Clone the Git repo. Do NOT download it as ZIP, that won't work.

   - Open a terminal or command prompt.
   - Run the following command:
     ```
     git clone https://github.com/Ahrimdon/xlabs.git
     ```

## Repository Contents
  - All XLabs Clients
  - All XLabs Dedicated Server Configuration Files
  - Complete XLabs Launcher
  - All Original Game Binaries
  - xlabs.dev Website
  - Client Documentation (e.g. Changelogs, README's, etc.)

## Installation

### Clients

1. Move the the folder contents of the desired client into your game folder.
2. Launch the executable.
   > **Note**: The executable and any accompanying folders should be in the base game directory.

### Launcher

1. Move the `xlabs` folder from `localappdata/xlabs` to your LocalAppData directory.
    > **Note**: On Windows, press `Win + R`, type `%localappdata%`, click ok.
2. Extract `IW4XLauncher.exe` and run it.
3. Select the proper game path for your game(s) of choice and launch.

## Disclaimer

This software has been created purely for the purposes of
academic research. It is not intended to be used to attack
other systems. Project maintainers are not responsible or
liable for misuse of the software. Use responsibly.


iw4x/CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog v0.3.0](http://keepachangelog.com/en/0.3.0/).

## r4303 - 2023-05-17

### Added

- Add `dumpZone` ZoneBuilder command (#1033)
- Add `FreezeControlsAllowLook` GSC method (#1026)

### Fixed

- The sound fix was removed as the bug no longer happens (#1007)
- Fix crash when the opening the "server info" menu on an empty server list (#1041)
- Fix patch that was causing the game thread to deadlock (#1038)

## r4251 - 2023-05-03

### Added

- Add GSC client field `ping` as a read-only field (#1002)
- Add GSC client field `address` as a read-only field (#1003)
- Add to the iw4x-rawfiles `common_scripts\utility` GSC script `getIP` function.

### Fixed

- `getPing` function in `common_scripts\utility` now works.

### Known issues

- Sound issue fix is experimental as the bug is not fully understood.

## r4246 - 2023-05-03

### Added

- Add to the iw4x-rawfiles `common_scripts\utility` GSC script `setPing` & `getPing` functions for backward compatibility.

### Fixed

- Fix bug with Steam Proxy (#991)
- Fix bug with the `say` GSC notify in regards to hidden chat messages (#989)

### Removed

- SetPing GSC method (#998)

### Known issues

- Sound issue fix is experimental as the bug is not fully understood.

## r4226 - 2023-04-26

### Changed

- Chat system will go back to using `SV_CMD_CAN_IGNORE` commands (#972)

### Security

- Check the address of the sender for the `print` OOB packet (#969)
- Check the address of the sender for the `voice` OOB packet (#973)

### Fixed

- Fix bug with how `sv_mapRotationCurrent` is parsed (#977)

### Known issues

- Sound issue fix is experimental as the bug is not fully understood.

## r4208 - 2023-04-22

### Changed

- `Noclip` GSC method does not require `sv_cheats` to be set to "1" for it to work (#962)
- `Ufo` GSC method does not require `sv_cheats` to be set to "1" for it to work (#962)

### Fixed

- Fix `InfoString` output (#961)
- Fix parsing of the server info (client-side) (#953)
- Fix bug in the /info TCP endpoint (#955)

### Known issues

- Sound issue fix is experimental as the bug is not fully understood.

## r4193 - 2023-04-19

### Added

- Add `LogString` GSC function (#895)
- Add `LogString` GSC method (#895)
- Add `sv_replaceTestClients` Dvar (#930)

### Changed

- `sv_mapRotationCurrent` supports `exec` directive for executing cfg scripts from the `game_settings` folder (#916)
- `SetPing` GSC method is now deprecated (#903)

### Security

- Fix DOS exploit in the network protocol of the game (#942)

### Fixed

- `sv_privatePassword` will work as intended (#908)
- Fix crash when loading bots.txt file (#927)

### Removed

- Remove `GetIp` GSC method (#903)
- Remove `GetPing` GSC method (#903)

### Known issues

- Sound issue fix is experimental as the bug is not fully understood.

## [0.7.9] - 2023-03-31

### Added

- Game scripts can be loaded from the `scripts/mp`, `scripts/mp/<map name>` and `scripts/mp/<game type>` folders (#859)
- Add `ReadStream` GSC function (#862)
- Add `IString` GSC function (#877)

### Changed

- Test Clients will no longer receive names from the Xlabs Patreon website. The old behaviour was restored (#852)
- Enabled `OpenFile` GSC function (#862)
- Enabled `CloseFile` GSC function (#862)
- Chat system will now use `SV_CMD_RELIABLE` commands to mitigate message duplication (#873)
- The built-in GSC compiler no longer throws fatal errors when overriding a built-in function or method (IW3 behaviour) (#880)
- `CastFloat` GSC function was renamed to `Float` (#880)

### Fixed

- Fix bug where knife lounges would not work with a gamepad (#848)
- Fix rare RCon crash (#861)
- Fix bug where `sv_RandomBotNames` stopped working (#876)
- Fix crash caused by "Mongoose" (dependency) (#874)

### Known issues

- Sound issue fix is experimental as the bug is not fully understood.

## [0.7.8] - 2023-03-17

### Added

- Clients can unprotect "saved" Dvars using the command line argument `-unprotect-dvars` (#694)
- Clients can protect all Dvars from being modified by the server using the command line argument `-protect-dvars` (#823)
- Add helpful information to script-related errors. Must set `developer` Dvar to "2" and `developer_script` Dvar to "1" (#721)
- Add command line `-disable-mongoose` argument to disable IW4x's built-in webserver (#728)
- Add command line `-disable-rate-limit-check` argument to disable the rate check on RCon requests (#769)
- GUID of muted clients is saved to the `muted-users.json` file in the `userraw` folder (#732)
- Add `sv_nextMap` Dvar (#736)
- Add `elifdef` and `elifndef` directives to the menu preprocessor (#747)
- Add `r_drawRunners` Dvar (#758)
- Add `r_drawLights` Dvar (#814)
- Add Discord Rich Presence (#761)
- Add `rcon_timeout` Dvar (#769)
- Add `sv_clanName` Dvar (#771)
- Add `InitialWeaponRaise` GSC method (#804)
- Add `RconWhitelistAdd` SV Command (#804)
- Add more options to `bg_bouncesAllAngles` Dvar (#806)
- Add `GetChar` GSC function (#813)
- Add `bg_bunnyHopAuto` Dvar (#818)
- Add `bg_disableLandingSlowdown` Dvar (#818)
- Add new map porting utility tool that makes the map porting process between CoD8 to CoD6 easy

### Changed

- Servers can no longer modify client Dvars that are saved in the client's config (#694)
- `banClient` and `muteClient` server commands do not apply to bots anymore (#730)
- The max value of `perk_extendedMeleeRange` Dvar was increased (#782)
- Test Clients will receive names from the Xlabs Patreon website in addition to the names from the bots.txt file (#771)

### Fixed

- Fix bug where`reloadmenus` command would not free resources used by custom menus (#740)
- Fix bug where demo playback would stop when opening a laptop based killstreak (#699)
- Fix bug where mod download speed was inexplicably slow (#707)
- Fix bug where `g_hardcore` could not be set by servers (#708)
- Fix "user cmd angles" for test clients (#707)

website/js/menu.js
(function ($) {
	"use strict";

	$(function () {
		var header = $(".start-style");
		$(window).scroll(function () {
			var scroll = $(window).scrollTop();

			if (scroll >= 10) {
				header.removeClass('start-style').addClass("scroll-on");
			} else {
				header.removeClass("scroll-on").addClass('start-style');
			}
		});
	});

	//Animation

	$(document).ready(function () {
		$('body.hero-anime').removeClass('hero-anime');
	});

	//Menu On Hover

	$('body').on('mouseenter mouseleave', '.nav-item', function (e) {
		if ($(window).width() > 750) {
			var _d = $(e.target).closest('.nav-item'); _d.addClass('show');
			setTimeout(function () {
				_d[_d.is(':hover') ? 'addClass' : 'removeClass']('show');
			}, 1);
		}}
	);
	
	//Switch light/dark

	$("#switch").on('click', function () {
		if ($("body").hasClass("dark")) {
			$("body").removeClass("dark");
			$("#switch").removeClass("switched");
		}
		else {
			$("body").addClass("dark");
			$("#switch").addClass("switched");
		}
	});
	


	// Cleanup Menus
	
	$(".iw4x-menu").on('click', function () {
		$(".iw6x-menu").removeClass("open");
		$(".iw6x-menu").removeClass("show");
		$(".guides").removeClass("open");
		$(".guides").removeClass("show");
		$(".s1x-menu").removeClass("open");
		$(".s1x-menu").removeClass("show");
		$(".socials").removeClass("open");
		$(".socials").removeClass("show");
	});
	
	$(".iw6x-menu").on('click', function () {
		$(".iw4x-menu").removeClass("open");
		$(".iw4x-menu").removeClass("show");
		$(".guides").removeClass("open");
		$(".guides").removeClass("show");
		$(".s1x-menu").removeClass("open");
		$(".s1x-menu").removeClass("show");
		$(".socials").removeClass("open");
		$(".socials").removeClass("show");
	});
	
	$(".s1x-menu").on('click', function () {
		$(".iw6x-menu").removeClass("open");
		$(".iw6x-menu").removeClass("show");
		$(".iw4x-menu").removeClass("open");
		$(".iw4x-menu").removeClass("show");
		$(".guides").removeClass("open");
		$(".guides").removeClass("show");
		$(".socials").removeClass("open");
		$(".socials").removeClass("show");
	});
	
	$(".guides").on('click', function () {
		$(".iw6x-menu").removeClass("open");
		$(".iw6x-menu").removeClass("show");
		$(".iw4x-menu").removeClass("open");
		$(".iw4x-menu").removeClass("show");
		$(".s1x-menu").removeClass("open");
		$(".s1x-menu").removeClass("show");
		$(".socials").removeClass("open");
		$(".socials").removeClass("show");
	});
	
	$(".socials").on('click', function () {
		$(".iw6x-menu").removeClass("open");
		$(".iw6x-menu").removeClass("show");
		$(".iw4x-menu").removeClass("open");
		$(".iw4x-menu").removeClass("show");
		$(".s1x-menu").removeClass("open");
		$(".s1x-menu").removeClass("show");
		$(".guides").removeClass("open");
		$(".guides").removeClass("show");
	});

	// Clean up hover menus

	$('.iw4x-menu').mouseover(function() {
		$(".iw6x-menu").removeClass("open");
		$(".iw6x-menu").removeClass("show");
		$(".guides").removeClass("open");
		$(".guides").removeClass("show");
		$(".s1x-menu").removeClass("open");
		$(".s1x-menu").removeClass("show");
		$(".socials").removeClass("open");
		$(".socials").removeClass("show");
	  });

	$('.iw6x-menu').mouseover(function() {
		$(".iw4x-menu").removeClass("open");
		$(".iw4x-menu").removeClass("show");
		$(".guides").removeClass("open");
		$(".guides").removeClass("show");
		$(".s1x-menu").removeClass("open");
		$(".s1x-menu").removeClass("show");
		$(".socials").removeClass("open");
		$(".socials").removeClass("show");
	  });
	  
	$('.s1x-menu').mouseover(function() {
		$(".iw4x-menu").removeClass("open");
		$(".iw4x-menu").removeClass("show");
		$(".iw6x-menu").removeClass("open");
		$(".iw6x-menu").removeClass("show");
		$(".guides").removeClass("open");
		$(".guides").removeClass("show");
		$(".socials").removeClass("open");
		$(".socials").removeClass("show");
	  });
	  
	$('.guides').mouseover(function() {
		$(".iw4x-menu").removeClass("open");
		$(".iw4x-menu").removeClass("show");
		$(".iw6x-menu").removeClass("open");
		$(".iw6x-menu").removeClass("show");
		$(".s1x-menu").removeClass("open");
		$(".s1x-menu").removeClass("show");
		$(".socials").removeClass("open");
		$(".socials").removeClass("show");
	  });
	  
	$('.socials').mouseover(function() {
		$(".iw4x-menu").removeClass("open");
		$(".iw4x-menu").removeClass("show");
		$(".iw6x-menu").removeClass("open");
		$(".iw6x-menu").removeClass("show");
		$(".s1x-menu").removeClass("open");
		$(".s1x-menu").removeClass("show");
		$(".guides").removeClass("open");
		$(".guides").removeClass("show");
	  });

})(jQuery); 

localappdata/xlabs/data/launcher-ui/js/main.js
window.addEventListener("load", initialize);
window.channel = window.executeCommand("get-channel");

function sleep(milliseconds) {
    return new Promise(resolve => {
        setTimeout(resolve, milliseconds);
    });
}

function makeSleep(milliseconds) {
    return () => sleep(milliseconds);
}

function waitForAllImages() {
    return new Promise(resolve => {
        function waitForAllImagesInternal() {
            const images = document.querySelectorAll('img');

            for (var i = 0; i < images.length; ++i) {
                if (!images[i].complete) {
                    window.requestAnimationFrame(waitForAllImagesInternal);
                    return;
                }
            }

            resolve();
        }

        waitForAllImagesInternal();
    });
}

function addStyleElement(css) {
    var head = document.head || document.getElementsByTagName('head')[0],
        style = document.createElement('style');

    head.appendChild(style);

    style.type = 'text/css';
    if (style.styleSheet) {
        // This is required for IE8 and below.
        style.styleSheet.cssText = css;
    } else {
        style.appendChild(document.createTextNode(css));
    }
}

function getOtherChannel(channel) {
    if (channel == "main") {
        return "dev";
    }
    return "main";
}

function adjustChannelElements() {
    window.channel.then(channel => {
        addStyleElement(`.channel-${getOtherChannel(channel)}{display: none;}`);
    });
}

function initialize() {
    initializeNavigation() //
        .then(() => waitForAllImages()) //
        .then(makeSleep(300))
        .then(() => window.executeCommand("show"));

    document.querySelector("#minimize-button").onclick = () => {
        window.executeCommand("minimize");
    };

    document.querySelector("#close-button").onclick = () => {
        window.executeCommand("close");
    };

    adjustChannelElements();
}

window.showSettings = function() {
    document.querySelector("#navigation>#settings.element").click();
}

function initializeNavigation() {
    var elements = document.querySelectorAll("#navigation>.element");
    elements.forEach(e => {
        e.addEventListener("click", handleNavigationClick);
    });

    return loadInitialPage();
}

function removeActiveElement() {
    var element = document.querySelector("#navigation>.element.active");
    if (element) {
        element.classList.remove("active");
    }
}

function handleNavigationClick(e) {
    const el = this;
    if (el.classList.contains("active")) {
        return;
    }

    removeActiveElement();
    el.classList.add("active");
    loadNavigationPage(el.id);
}

function loadInitialPage() {
    const el = document.querySelector("#navigation>.element.active");
    return loadNavigationPage(el.id);
}

function setInnerHTML(elm, html) {
    elm.innerHTML = html;
    Array.from(elm.querySelectorAll("script")).forEach(oldScript => {
        const newScript = document.createElement("script");
        Array.from(oldScript.attributes)
            .forEach(attr => newScript.setAttribute(attr.name, attr.value));
        newScript.appendChild(document.createTextNode(oldScript.innerHTML));
        oldScript.parentNode.replaceChild(newScript, oldScript);
    });
}

function loadNavigationPage(page) {
    var content = document.querySelector("#content");
    return fetch(`./pages/${page}.html`).then(data => {
        return data.text()
    }).then(text => {
        setInnerHTML(content, text);
    });
}

dedicated-server-configs/S1xServerConfigs/s1x/scripts/bots.lua
players = {}

function tablefind(tab,el)
	for index, value in pairs(tab) do
		if value == el then
			return index
		end
	end

	return nil
end

function countBots()
	local count = 0

	for index, value in pairs(players) do
		local isBot = game:isbot(value)
		if isBot == 1 then
			count = count + 1
		end
	end

	return count
end
function countPlayers()
	local count = 0

	for index, value in pairs(players) do
		local isBot = game:isbot(value)
		if isBot == 0 then
			count = count + 1
		end
	end

	return count
end

function removePlayer(player)
	local index = tablefind(players, player)
	if index ~= nil then
		table.remove(players, index)
	end
end

function kickPlayer(player)
	local clientNum = player:getentitynumber()
	game:kickplayer(clientNum)
end

function kickLeftoverBot(player)
  local timer = game:ontimeout(function() kickPlayer(player) end, 10000)
  local timer2 = player:onnotifyonce("spawned_player", function() timer:clear() end)
  player:onnotifyonce("disconnect", function() timer2:clear() end)
end

function player_connected(player)
	player:onnotifyonce("disconnect", function() removePlayer(player) end)
	table.insert(players, player)

	local isBot = game:isbot(player)
	if isBot == 1 then
		player:botsetdifficulty("default")
		kickLeftoverBot(player)
	end
end

function kickBot()
	for index, player in pairs(players) do
		local isBot = game:isbot(player)
		if isBot == 1 then
			kickPlayer(player)
			break
		end
	end
end

function spawnBot()
	game:executecommand("spawnbot 1");
end

function monitor()
	local count = #players;
	local botCount = countBots()
	if count < 10 and botCount < 5 then
		spawnBot()
	elseif count > 14 and botCount > 0 then
		kickBot()
	end
end

function startLogic()
	game:oninterval(monitor, 3000)
end

local bots = 6

if game:getdvar("gamemode") == "mp" then
    game:executecommand("spawnbot " .. bots);   
    level:onnotify("matchStartTimer", startLogic)
	level:onnotify("connected", player_connected)
end

dedicated-server-configs/S1xServerConfigs/s1x/scripts/utilities.lua
--[[ Utilities ]]--

function entity:setPoints( points )
    if points < 0 then
        points = 0
    end
    self.extrascore0 = points
    self.pers.extrascore0 = points
    self.score = points
    self.pers.score = points
end

function round(num, numDecimalPlaces)
	local mult = 10^(numDecimalPlaces or 0)
	return math.floor(num * mult + 0.5) / mult
end

function vector3ToString( v, numDecimalPlaces)
	if v[0] ~= nil and v[1] ~= nil and v[2] ~= nil then
		return ("(".. round(v[0],numDecimalPlaces) ..",".. round(v[1],numDecimalPlaces) ..",".. round(v[2],numDecimalPlaces) ..")");
	elseif v.x ~= nil and v.y ~= nil and v.z ~= nil then
		return ("(".. round(v.x, numDecimalPlaces) ..",".. round(v.y,numDecimalPlaces) ..",".. round(v.z,numDecimalPlaces) ..")");
	else
		return "Not valid vector";
	end
end

function split(pString, pPattern)
    Table = {}  -- NOTE: use {n = 0} in Lua-5.0
    fpat = "(.-)" .. pPattern
    last_end = 1
    s, e, cap = pString:find(fpat, 1)
    while s do
       if s ~= 1 or cap ~= "" then
      table.insert(Table,cap)
       end
       last_end = e+1
       s, e, cap = pString:find(fpat, last_end)
    end
    if last_end <= #pString then
       cap = pString:sub(last_end)
       table.insert(Table, cap)
    end
    return Table
 end

 function indexof(t, value)
	for i = 1, #t do
		if (value == t[i]) then
			return i
		end
	end

	return -1
end

function size(T)
    count = 0
    for _ in pairs(T) do count = count + 1 end
    return count
  end

dedicated-server-configs/S1xServerConfigs/s1x/scripts/__init__.lua
--[[
   Mod: Custom GunGame
   Version: V1.0.0
   Client: s1x - COD Advanced Warfare
   Author: DoktorSAS 
   Special Thanks to quaK

   Socials:
   Twitter  : https://twitter.com/DoktorSAS
   Discord  : https://discord.gg/8HNGUTVvmd
   Youtube  : https://www.youtube.com/channel/UCfMGaICm89h7g-54iBTFA4w
   Github   : https://github.com/DoktorSAS
]]--

--[[ Gametype Config ]]--
local Config = {}
--[[
    1 -> Enable
    0 -> Disable
]]--
Config.HighJump = 1
Config.HighJumpDrop = 1
Config.Dodge = 1
Config.PowerSlide = 1
Config.BoostJump = 1
Config.KillstreaksPerks = 1
Config.HumiliatonsForKill = 1
--[[
    dm  -> Free For All
    war -> Team Deathmatch
]]--
Config.Gametype = "war"

--[[ GunGame ]]--
include("utilities")
if game:getdvar("g_gametype") ~= Config.Gametype then
    game:setdvar("g_gametype", Config.Gametype)
	game:executecommand("map_restart")
end
game:setdvar("sct_game_high_jump", Config.HighJump)
if Config.Gametype == "dm" then
    game:setdvar("scr_dm_score_kill", 1)
    game:setdvar("scr_dm_score_assist", 0)
elseif Config.Gametype == "war" then
    game:setdvar("scr_war_score_kill", 1)
    game:setdvar("scr_war_score_assist", 0)
end

game:setdvar("g_maxDroppedWeapons", 2)

require("bots")

local colors = {    red = "^1",
					green = "^2",
					yellow = "^3",
					blue = "^4",
					cyan = "^5",
					pink = "^6",
                    white = "^7",
					team_color = "^8",
					grey = "^9",
					black = "^0",
					rainbow = "^:"
					}
local teams = { 	spectator = "spectator",
					axis = "axis",
					allies = "allies"
					}
local sessionstates = {	spectator = "spectator",
						dead = "dead",
						playing = "playing"
}

local social = {    author = "DoktorSAS",
                    twitter = "DoktorSAS",
					instagram = "DoktorSAS",
					website = "https://github.com/DoktorSAS",
					discord = "https://discord.gg/8HNGUTVvmd"
}

print("=====================================================")
print("\t GunGame Developed by " .. social.author)
print("\t Special thanks to quaK")
print("=====================================================")
local players = {}
for i = 0, 18, 1 do
    players[i] = {isBot = false, suicide = false, gun = 1, humiliator = "", lastgun = "", streak = 0, isConnected = false};
end

local guns = {
                "iw5_pbw_mp", -- Pistol
                "iw5_rw1_mp",
                "iw5_titan45_mp",
                "iw5_uts19_mp", -- Shotgun
                "iw5_maul_mp",
                "iw5_rhino_mp",
                "iw5_kf5_mp", -- SMG
                "iw5_sac3_mp_akimbosac3",
                "iw5_asm1_mp",
                "iw5_hmr9_mp",
                "iw5_sn6_mp",
                "iw5_mp11_mp",
                "iw5_himar_mp",
                "iw5_hbra3_mp", -- Rifle
                "iw5_bal27_mp",
                "iw5_ak12_mp",
                "iw5_epm3_mp",
                "iw5_arx160_mp",
                "iw5_m182spr_mp",
                "iw5_asaw_mp", -- MG
                "iw5_exoxmg_mp",
                "iw5_lsat_mp",
                "iw5_mahem_mp", -- Launcher
                "iw5_maaws_mp",
                "iw5_mors_mp", -- Snipers
                "iw5_thor_mp",
                "iw5_m990_mp",
                "iw5_gm6_mp",
                "iw5_em1_mp", -- Beam
                "iw5_exocrossbow_mp", -- Crossbow
                "iw5_combatknife_mp" -- Knife
}
if Config.Gametype == "war" then
    game:setdvar("scr_war_scorelimit", 0) 
elseif Config.Gametype == "dm" then
    game:setdvar("scr_dm_scorelimit", size(guns)) 
end

function entity:onPlayerSpawned(  )
    entity_number = self:getentitynumber()
    if players[entity_number].sucide then
        self:iclientprintlnbold("You got ".. colors.red .. "Humiliated")
        players[entity_number].sucide = false
        if players[entity_number].gun > 1 then
            players[entity_number].gun = players[entity_number].gun - 1
        end
        self:setPoints( (indexof(guns, guns[players[entity_number].gun])-1+Config.HumiliatonsForKill) )
        self:playlocalsound("mp_prestige_purchase") 
    elseif players[entity_number].humiliator ~= "" then
        self:iclientprintlnbold("You got ".. colors.red .. "Humiliated")
        self:iclientprintln(players[entity_number].humiliator .." humiliate " .. colors.white .. "you")
        players[entity_number].humiliator = ""
        if players[entity_number].gun > 1 then
            players[entity_number].gun = players[entity_number].gun - 1
        end
        self:setPoints( (indexof(guns, guns[players[entity_number].gun])-1+Config.HumiliatonsForKill) )
        self:playlocalsound("mp_prestige_purchase")
    end
    self:takeallweapons()
    if players[entity_number].gun <= 0 then
        players[entity_number].gun = 1
    end
    self:giveweapon(guns[players[entity_number].gun])
    self:switchtoweapon(guns[players[entity_number].gun])
    self:setspawnweapon(guns[players[entity_number].gun])
    game:ontimeout(function()
        self:clearperks()
    end, 50)
end

function entity:onWeaponChange(  )
    entity_number = self:getentitynumber()
    if self:getcurrentweapon() ~= "none" and self:getcurrentweapon() ~= nil and self:getcurrentweapon() ~= guns[players[entity_number].gun] then
        self:takeweapon( self:getcurrentweapon() )
        if players[entity_number].gun <= 0 then
            players[entity_number].gun = 1
        end
        if self.score > 0 and self:hasweapon( guns[players[entity_number].gun] ) == 1 then 
            self:switchtoweapon(guns[players[entity_number].gun])
        else
            self:giveweapon(guns[players[entity_number].gun])
            self:switchtoweapon(guns[players[entity_number].gun])
        end
    else
        self:switchtoweapon(guns[players[entity_number].gun])
    end
end

function forceend()
    if game:getdvar("g_gametype") == "war" then
        game:setdvar("scr_war_timelimit", 0.1)
        game:executecommand("set scr_war_timelimit 0.01")
    end
end

function entity:onWeaponReload(  )
    if( players[entity_number].humiliator ~= "") then
        players[entity_number].humiliator = ""
    end
	self:givestartammo( self:getcurrentweapon() )
    
end


local bots_spawned = 8
local players_in_game = 0

local onPlayerConnected = function ( player )

    local onMovements = game:oninterval(function ()
        player:allowdodge(Config.Dodge)
        player:allowpowerslide(Config.PowerSlide)
        player:allowboostjump(Config.BoostJump)
        player:allowhighjump(Config.HighJump)
        player:allowhighjumpdrop(Config.HighJumpDrop)
    end, 50)

    if string.find(player:getguid(), "bot") then
        players[player:getentitynumber()].isBot = true
    else
        players[player:getentitynumber()].isBot = false
    end
    local pID = player:getentitynumber();
    players_in_game = players_in_game + 1
    players[player:getentitynumber()].gun = 1
    players[player:getentitynumber()].isConnected = true

    local onPlayerSpawned = player:onnotify("spawned_player", function()
        player:iclientprintln("Mod developed by @"..colors.cyan.. social.twitter);
        player:iclientprintln( colors.red .. "kill".. colors.white .." players to progress with weapons");
        player:iclientprintln("Welcome to".. colors.cyan .." GunGame " .. colors.white .. "Server");
        player:onPlayerSpawned( )
    end)
    local onWeaponChange = player:onnotify("weapon_change", function()
        player:onWeaponChange()
    end)
    local onPlayerReload = player:onnotify("reload", function( ) 
        player:onWeaponReload()
    end)

    local onWeaponFired = player:onnotify("weapon_fired", function ( weapon )
        if weapon == "iw5_maaws_mp" or weapon == "throwingknife_mp" then
            player:givemaxammo( player:getcurrentweapon() )
        end
    end)
    
    local onGivingLoadout = player:onnotify("givingLoadout", function ( )
        player:onPlayerSpawned()
    end)
    
    player:onnotifyonce("disconnect", function ()
        onPlayerSpawned:clear()
        onWeaponChange:clear()
        onPlayerReload:clear()
        onWeaponFired:clear()
        onGivingLoadout:clear()
        onMovements:clear()
        players[pID] = {isBot = false, suicide = false, gun = 1, humiliator = "", lastgun = "", streak = 0, isConnected = false};
        players_in_game = players_in_game - 1
    end);
end

function show_message(player, msg)
    local elem = game:newclienthudelem(player)
    elem.elemType = "font"
    elem.font = "default"
    elem.fontscale = 2
    elem.x = 0
    elem.y = -100
    elem.alignx = "center"
    elem.aligny = "middle"
    elem.horzalign = "center"
    elem.vertalign = "middle"
    elem.color = vector:new(1.0, 1.0, 1.0)
    elem.alpha = 1
    elem:settext( msg )
    elem:setpulsefx(40, 6000, 600)
    
    game:ontimeout(function()
        elem:destroy()
    end, 8000)
end

function team_win( team )
    local players = game:getentarray("player", "classname")
    for i = 1, #players do 
        if players[i].team == team then
            show_message(players[i], "Your team ^2Win ^7the Game")
        else
            show_message(players[i], "Your team ^1Lost ^7the Game")
        end
    end
   
end


level:onnotifyonce("connected", function ()

    game:onplayerdamage(function(_self, _inflictor, _attacker, damage, dflags, _mod, _weapon, _vPoint, _vDir, _hitLoc)
        return damage
    end)

    game:onplayerkilled(function(_self, _inflictor, _attacker, damage, _mod, _weapon, _vPoint, _vDir, _hitLoc)
        if _attacker.name ~= nil and _attacker:getentitynumber() < 18 then
            entity_number = _attacker:getentitynumber()
            if  players[entity_number].humiliator ~= "" then
                players[entity_number].humiliator = ""
            end
            game_ended = false
            if _mod == "MOD_MELEE_ALT" then
                players[_self:getentitynumber()].humiliator = _attacker.name
                _attacker:setPoints( (_attacker.score-1) )
            elseif _self:getguid() == _attacker:getguid() then
                players[entity_number].humiliator = ""
                _self:setPoints( (_self.score-1+Config.HumiliatonsForKill) )
            elseif players[entity_number].gun > 0 and _weapon == guns[players[entity_number].gun] and (players[entity_number].lastgun == "" or players[entity_number].lastgun ~= guns[players[entity_number].gun]) then
                if Config.Gametype == "war" then
                    if players[entity_number].gun >= size(guns) then
                        forceend()
                        team_win( _attacker.team )
                        game_ended = true
                    end
                end
                if game_ended == false then
                    players[entity_number].lastgun = _weapon
                    _attacker:takeweapon(guns[players[entity_number].gun])
                    players[entity_number].lastgun = guns[players[entity_number].gun-1] 
                    players[entity_number].gun = players[entity_number].gun + 1
                    _attacker:giveweapon(guns[players[entity_number].gun]);
                    _attacker:switchtoweapon(guns[players[entity_number].gun])
                    _attacker:playlocalsound("mp_level_up")
                    players[entity_number].streak = players[entity_number].streak + 1
                    if Config.KillstreaksPerks then
                        if players[entity_number].streak == 3 then
                            _attacker:iclientprintlnbold("Quick Reload ^6Unlocked")
                            _attacker:setperk("specialty_quickdraw", 1, 1)
                        elseif players[entity_number].streak == 5 then
                            _attacker:iclientprintlnbold("Sprint Reload ^6Unlocked")
                            _attacker:setperk("specialty_sprintreload", 1, 1)
                        elseif players[entity_number].streak == 8 then
                            _attacker:iclientprintlnbold("Longer Sprint ^6Unlocked")
                            _attacker:setperk("specialty_longersprint", 1, 1)
                        elseif players[entity_number].streak == 11 then
                            _attacker:iclientprintlnbold("Faster Melee ^6Unlocked")
                            _attacker:setperk("specialty_fastermelee", 1, 1)
                        elseif players[entity_number].streak == 14 then
                            _attacker:iclientprintlnbold("Faster Sprint Recovery ^6Unlocked")
                            _attacker:setperk("specialty_fastsprintrecovery", 1, 1)
                        elseif players[entity_number].streak == 17 then
                            _attacker:iclientprintlnbold("Steady Aim ^6Unlocked")
                            _attacker:setperk("specialty_steadyaimpro", 1, 1)
                        end  
                    end
                end
            elseif players[entity_number].lastgun == "" then
                players[entity_number].lastgun = _weapon
            end
            players[_self:getentitynumber()].streak = 0; 
            _attacker:setPoints( (indexof(guns, guns[players[entity_number].gun])-2) )
            _attacker:setnormalhealth(_attacker.maxhealth )
          
           
        end
        if _weapon ~= nil and _weapon ~= "" and _mod == "MOD_TRIGGER_HURT" or _mod == "MOD_SUICIDE" or _mod == "MOD_PROJECTILE_SPLASH" or  _mod == "MOD_TRIGGER_HURT" then
            players[_self:getentitynumber()].humiliator = _self.name
        elseif _mod == "MOD_TRIGGER_HURT" or _mod == "MOD_SUICIDE" or _mod == "MOD_PROJECTILE_SPLASH" or  _mod == "MOD_TRIGGER_HURT" then
            players[_self:getentitynumber()].sucide = true
            _self:setPoints( (_self.score-1+Config.HumiliatonsForKill) )
        end
    end)
    
end);

level:onnotify("connected", onPlayerConnected)